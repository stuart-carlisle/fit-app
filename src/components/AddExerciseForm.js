import React, { useState , useEffect} from 'react'
import moment from 'moment'
import { connect } from 'react-redux'
import numeral from 'numeral'

export const AddExerciseForm = ({
  filters = {},
  onSubmit, 
  exercise = {
    description:'',
    type:'',
    energy:'',
    time:''
  }
}) => {
    const defaultEnergy = numeral(exercise.energy).value()
    const defaultTime = numeral(exercise.time).value()
    const date = filters.date
    const referenceDatabase = filters.referenceDatabase
    const [time, setTime] = useState(defaultTime)
    const [error, setError] = useState('')
    const [energy, setEnergy] = useState(defaultEnergy)
    const diaryDate = date ? date.valueOf() : moment().startOf('day').valueOf()
    const lastUsed = moment().valueOf()

    useEffect(()=>{
        
      let refreshed = JSON.parse(localStorage.getItem('refreshed'))
      if(refreshed){
        window.history.back()
        window.scrollTo(0,0)
        localStorage.removeItem('refreshed')
      }else{
        localStorage.setItem('refreshed',JSON.stringify(true))
      }
    },[])

    const onFormSubmit = (e) => {
        e.preventDefault();
        const exerciseData = { 
          description: exercise.description,
          type: exercise.type
        }
        if (numeral(time).value()<=0) {
          setError('Please provide a time greater than 0')
          window.scrollTo(0,0)
        } else {
          setError('')
          onSubmit(
            referenceDatabase,
            diaryDate,
            lastUsed,
            exerciseData,
            time.toString(),
            energy.toString()
          )
          localStorage.removeItem('refreshed')
        }
      }
    
    const onTimeChange = (e)=>{
        const time = e.target.value
          if (!time || time.match(/^\d{1,}(\.\d{0,1})?$/)) {
          setTime(time)
          const timeValue = numeral(time).value()
          setEnergy(Math.round(defaultEnergy * (timeValue/defaultTime)))
        }
    }

    const onEnergyChange = ()=>{
      
    }

    return (
        <div>
        <div className="content-container__background">
        <div className="content-container__form__update">
            <form className= "form" onSubmit={onFormSubmit}>
                {error && <p id="error-message" className="form__error">{error}</p>}
                  <h1 className="page-header__title-update"> {exercise.description} </h1>
                    <span className="text-input__span-container-form">
                      <div className="content-container__update-titles">
                        <h3 className="form__heading__update">Exercise Time (mins):</h3>
                        <h3 className="form__heading__update">Energy (calories): </h3>
                      </div>
                      <div className="content-container__update">
                        <input  //Exercise Duration
                          className="text-input__update-form"
                          type="text"
                          value={time}
                          onChange={onTimeChange}
                          placeholder="Enter a Duration"
                        />  
                        <input  //Energy Used
                          className="text-input__update-form"
                          type="text"
                          value={energy?energy:''}
                          onChange={onEnergyChange}
                          placeholder="Energy Used"
                        /> 
                      </div>
                    </span>
                <button className="button button--form">Add Exercise</button>
            </form>
          </div>
        </div>
      </div>  
    )
}

const mapStateToProps = (state) => {
  return {
    filters: state.filters
  }
}


export default connect(mapStateToProps)(AddExerciseForm);