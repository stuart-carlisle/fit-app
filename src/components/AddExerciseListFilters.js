import React, { useState } from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByDate, sortByAlphabetical, setReferenceDatabaseToPublic } from '../actions/filters';
import { setExercisesPublic } from '../actions/exercises'
import algoliasearch from 'algoliasearch/lite'

export const AddExerciseListFilters = ({
    filters,
    setTextFilter, 
    setReferenceDatabasePublic,
    sortByDate, 
    sortByAlphabetical,
    setExercisesPublic 
  }) => {
 
  const searchClient = algoliasearch (
      process.env.ALGOLIA_APP_ID,
      process.env.ALGOLIA_API_KEY
  )

  const index = searchClient.initIndex('exercises');

  const [text, setText] = useState('')
    
  const onTextChange = (e) => {
    setTextFilter(e.target.value)
    setText(e.target.value)
  }

  const onSortChange = (e) => {
    if (e.target.value === 'date') {
      sortByDate()
    } else if (e.target.value === 'alphabetical') {
      sortByAlphabetical()
    }
  }

  const onSubmit = (e) => {
    e.preventDefault()
    index
      .search({
        query: text
      })
      .then((responses) => {
        const array = responses.hits
        const newExercises = array.map((exercise)=>{
          return{
            description:exercise.description,
            type:exercise.type,
            energy:exercise.energy,
            time:exercise.time,
            id:exercise.objectID
          }
        })
        setExercisesPublic(newExercises)
      }).then(()=>{
        setReferenceDatabasePublic()
      })
}

    return (
      
      <div className="header__content--filters"> 
      <form className="filters-text-input__span-container" onSubmit={onSubmit}>
          <div className="input-group__item-page-header">
            <input
              type="text"
              className="filters-text-input"
              placeholder="SEARCH EXERCISES"
              value={text}
              onChange={onTextChange}
            />
          </div>
          <div className="input-group__item-page-header">
            <select
              className="minimal"
              value={filters.sortBy}
              onChange={onSortChange}
            >
              <option value="date">DATE</option>
              <option value="alphabetical">ALPHABETICAL</option>
            </select>
          </div>
            <button className="button filters-button--form-page-header">Search Online</button>
            </form>   
        </div>
    )
}

const mapStateToProps = (state) => {
  return {
    filters: state.filters
  };
};

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAlphabetical: () => dispatch(sortByAlphabetical()),
  setReferenceDatabasePublic: () => dispatch(setReferenceDatabaseToPublic()),
  setExercisesPublic: (exercises) => dispatch(setExercisesPublic(exercises))
})

export default connect(mapStateToProps,mapDispatchToProps)(AddExerciseListFilters);

