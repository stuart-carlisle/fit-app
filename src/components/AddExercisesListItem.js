import React from 'react';
import { Link } from 'react-router-dom';

export const AddExercisesListItem = ({ description, energy, id }) => (


  <Link className="list-item" to={`/add-exercise-confirmation/${id}`}>
    <div>
      <p className="list-item__title">{description}</p>
    </div>
    <div>
      <p className="list-item__data">{energy}</p>
    </div>  
  </Link>

);

export default AddExercisesListItem;
