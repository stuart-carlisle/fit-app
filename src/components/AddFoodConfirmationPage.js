import React from 'react'
import { connect } from 'react-redux'
import AddFoodForm from './AddFoodForm'
import { Link } from 'react-router-dom'
import { startAddFoodDiary, startAddFoodPrivate, startUpdateFoodPrivate } from '../actions/foods'

export const AddFoodConfirmationPage = (props) => {
  
  const onSubmit = (referenceDatabase, diaryDate, lastUsed, numberOfServings, food, meal) => {
    if (referenceDatabase==='public'){
      props.startAddFoodPrivate({...food,lastUsed})
    } else if (referenceDatabase==='private'){
      props.startUpdateFoodPrivate(props.food.id,{...food,lastUsed})
    }
    props.startAddFoodDiary({...food, diaryDate, numberOfServings, meal})
    props.history.push('/diary');
  }
  

  return (
     <div>
     <Link className="button__back-button" to={`/add`}><img className="image__header" src="/images/back-arrow.svg"/></Link>
        <div className="page-header__filter page-header__filter--with-back">
          <div className="content-container">
            <h1 className="page-header__title">ADD FOOD</h1>
          </div>
        </div>
        <div className="show-for-mobile page-header__margin"></div>
        <div className="content-container">
        <AddFoodForm id="add-food-form" targets={props.targets} food={props.food} onSubmit={onSubmit} />
        </div>
     </div>
 )
}

const mapStateToProps = (state,props) => {
  if(state.filters.referenceDatabase==='public'){
    return {
      food: state.foodsPublic.find((food) => food.id === props.match.params.id),
      targets: state.targets
    }
  }else if(state.filters.referenceDatabase==='private'){
    return {
      food: state.foodsPrivate.find((food) => food.id === props.match.params.id),
      targets: state.targets
    }
  }
}

const mapDispatchToProps = (dispatch) =>({
    startAddFoodDiary: (food) => dispatch(startAddFoodDiary(food)),
    startUpdateFoodPrivate: (id, food) => dispatch(startUpdateFoodPrivate(id, food)),
    startAddFoodPrivate: (food) => dispatch(startAddFoodPrivate(food))
})

export default connect(mapStateToProps, mapDispatchToProps)(AddFoodConfirmationPage);




