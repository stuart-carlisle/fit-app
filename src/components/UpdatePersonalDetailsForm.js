import React, { useState } from 'react'
import moment from 'moment'
import { SingleDatePicker } from 'react-dates'

export const UpdatePersonalDetailsForm = ({ onSubmit }) => {
    const [weight, setWeight] = useState('')
    const [date, setDate] = useState(moment().startOf('day'))
    const [error, setError] = useState('')
    const [calendarFocused, setCalendarFocused] = useState(null)

    const onFormSubmit = (e) => {
        e.preventDefault();
        if (weight.length===0) {
          setError('Enter a valid weight value')
        } else {
          setError('')
          onSubmit(
            weight,
            date
          )
        }
      }

    const onWeightChange = (e)=>{
        const weight = e.target.value
        if (!weight || weight.match(/^\d{1,}(\.\d{0,1})?$/)) {
            setWeight(weight)
        }
    }

    const onDateChange = (date) => {
      setDate(date)
    };

    const onFocusChange = ({focused}) => {
      setCalendarFocused(focused)
    }

    return (
        <div>
        <div className="content-container__background">
        <div className="content-container__form">
            <form className= "form" onSubmit={onFormSubmit}>
                {error && <p id="error-message" className="form__error">{error}</p>}
                <span className="text-input__span-container">
                        <input  //Weight
                            className="text-input"
                            type="text"
                            value={weight}
                            onChange={onWeightChange}
                            placeholder='Enter Weight in kg'
                        />
                        <div className="input-group__item input-group__item--date">
                          <SingleDatePicker
                            date={date}
                            onDateChange={onDateChange}
                            focused={calendarFocused}
                            onFocusChange={onFocusChange}
                            numberOfMonths={1}
                            isOutsideRange={() => false}
                            id="date_picker_weight"
                            withPortal={true}
                          />
                      </div>
                  </span>
                <button className="button button--form">Update Weight</button>
            </form>
        </div>
      </div> 
      </div>  
    )
}

export default UpdatePersonalDetailsForm
