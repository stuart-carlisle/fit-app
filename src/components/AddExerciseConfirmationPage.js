import React from 'react'
import { connect } from 'react-redux'
import AddExerciseForm from './AddExerciseForm'
import { Link } from 'react-router-dom'
import { startAddExerciseDiary, startAddExercisePrivate, startUpdateExercisePrivate } from '../actions/exercises'

export const AddExerciseConfirmationPage = (props) => {
  
  const onSubmit = (referenceDatabase, diaryDate, lastUsed, exerciseDescAndType, time, energy) => {
    if (referenceDatabase==='public'){
      props.startAddExercisePrivate({...exerciseDescAndType, time, energy, lastUsed})
    } else if (referenceDatabase==='private'){
      props.startUpdateExercisePrivate(props.exercise.id,{...exerciseDescAndType, time, energy, lastUsed})
    }
    props.startAddExerciseDiary({...exerciseDescAndType, time, diaryDate, energy})
    props.history.push('/diary');
  }
  

  return (
     <div>
     <Link className="button__back-button" to={`/add-exercise`}><img className="image__header" src="/images/back-arrow.svg"/></Link>
        <div className="page-header__filter page-header__filter--with-back">
          <div className="content-container">
            <h1 className="page-header__title">ADD EXERCISE</h1>
          </div>
        </div>
        <div className="show-for-mobile page-header__margin"></div>
        <div className="content-container">
        <AddExerciseForm id="add-exercise-form" exercise={props.exercise} onSubmit={onSubmit} />
        </div>
     </div>
 )
}

const mapStateToProps = (state,props) => {
  if(state.filters.referenceDatabase==='public'){
    return {
      exercise: state.exercisesPublic.find((exercise) => exercise.id === props.match.params.id)
    }
  }else if(state.filters.referenceDatabase==='private'){
    return {
      exercise: state.exercisesPrivate.find((exercise) => exercise.id === props.match.params.id)
    }
  }
}

const mapDispatchToProps = (dispatch) =>({
    startAddExerciseDiary: (exercise) => dispatch(startAddExerciseDiary(exercise)),
    startUpdateExercisePrivate: (id, exercise) => dispatch(startUpdateExercisePrivate(id, exercise)),
    startAddExercisePrivate: (exercise) => dispatch(startAddExercisePrivate(exercise))
})

export default connect(mapStateToProps, mapDispatchToProps)(AddExerciseConfirmationPage);




