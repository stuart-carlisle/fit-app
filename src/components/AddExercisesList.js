import React from 'react'
import { connect } from 'react-redux'
import AddExercisesListItem from './AddExercisesListItem'
import selectExercises from '../selectors/exercises'

export const AddExercisesList = ({exercises}) => {
  let exercisesTopResults
  
  if(exercises.length>10){
    exercisesTopResults = exercises.slice(0,10)
  }else{
    exercisesTopResults = exercises
  }

  return(<div className="content-container">
   <div className="list-header">
    <div className="">Exercise</div>
    <div className="show-for-tablet">Calories Burned</div>
    <div className="show-for-mobile">Cals Burned</div>
   </div>
    <div className="list-body">
      {
        exercisesTopResults.length === 0 ? (
          <div className="list-item list-item--message">
            <span>No Exercises In History</span>
          </div>
        ) : (
         exercisesTopResults.map((exercise) => {
            return <AddExercisesListItem key={exercise.id} {...exercise} />
          })
        )
      }
    </div>
  </div>  
  )
}

const mapStateToProps = (state) => {
  return {
    exercises: state.filters.referenceDatabase === 'public' ? state.exercisesPublic : selectExercises(state.exercisesPrivate,  state.filters)
  }
}

export default connect(mapStateToProps)(AddExercisesList);
