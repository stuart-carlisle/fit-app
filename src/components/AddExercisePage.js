import React, { useEffect } from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import AddExerciseListFilters from './AddExerciseListFilters'
import AddExercisesList from './AddExercisesList'
import { setTextFilter, sortByDate, setReferenceDatabaseToPrivate } from '../actions/filters'

export const AddExercisePage = ({ setTextFilter, sortByDate, setReferenceDatabaseToPrivate}) => {
  useEffect(()=>{
    localStorage.removeItem('refreshed')
      setTextFilter('')
      sortByDate()
      setReferenceDatabaseToPrivate()
  },[])
    return(
      <div>
      <Link className="button__back-button" to={`/diary`}><img className="image__header" src="/images/back-arrow.svg"/></Link>
    <div className="page-header__filter page-header__filter--with-back">
      <div className="content-container">
        <AddExerciseListFilters />
      </div>  
    </div>
    <AddExercisesList />
    <div className="content-container">
    <Link className="button button--link" to={`/create-exercise`}>+ CREATE CUSTOM EXERCISE</Link>
    </div>
    </div>
    )
}

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  setReferenceDatabaseToPrivate: () => dispatch(setReferenceDatabaseToPrivate())
})

export default connect (undefined, mapDispatchToProps)(AddExercisePage)