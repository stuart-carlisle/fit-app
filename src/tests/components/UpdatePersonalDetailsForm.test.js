import React from 'react'
import { shallow } from 'enzyme'
import { UpdatePersonalDetailsForm  } from '../../components/UpdatePersonalDetailsForm'
import  moment from 'moment'

let wrapper, onSubmit

beforeEach(()=>{
    onSubmit=jest.fn()
    wrapper = shallow(<UpdatePersonalDetailsForm onSubmit={onSubmit}/>)
})

test('Should render update personal details form correctly',()=>{
    expect(wrapper).toMatchSnapshot()
})

test('Should set weight on weight change',()=>{
    const value='94'
    wrapper.find('input').simulate('change',{
                target: { value }
            })
    expect(wrapper.find('input').prop('value')).toBe(value)
})

test('Should set weight on invalid weight change',()=>{
    const value='94.478'
    wrapper.find('input').simulate('change',{
                target: { value }
            })
    expect(wrapper.find('input').prop('value')).toBe('')
})

test('Should set date correctly on date change',()=>{
    const now = moment().add(2,'days')
    wrapper.find('withStyles(SingleDatePicker)').prop('onDateChange')(now)
    expect(wrapper.find('withStyles(SingleDatePicker)').prop('date')).toEqual(now)
})

test('Should set calendar on change',()=>{
    const focused = true
    wrapper.find('withStyles(SingleDatePicker)').prop('onFocusChange')({focused})
    expect(wrapper.find('withStyles(SingleDatePicker)').prop('focused')).toEqual(focused)
})

test('Should render error for invalid form submission',()=>{
    wrapper.find('form').simulate('submit',{
        preventDefault:() => { }
    })
    expect(wrapper.exists('#error-message')).toBe(true);
})

test('Should simulate form completion and call onSubmit prop for valid form submission',()=>{
    const value='94'
    wrapper.find('input').simulate('change',{
        target: { value: value }
    })
    const now = moment().add(2,'days')
    wrapper.find('withStyles(SingleDatePicker)').prop('onDateChange')(now)
    wrapper.find('form').simulate('submit',{
        preventDefault:() => { }
    })
    expect(wrapper.exists('#error-message')).toBe(false);
    expect(onSubmit).toHaveBeenLastCalledWith(value,now)
})

