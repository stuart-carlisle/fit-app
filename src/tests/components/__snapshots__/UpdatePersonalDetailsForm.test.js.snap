// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render update personal details form correctly 1`] = `
<div>
  <div
    className="content-container__background"
  >
    <div
      className="content-container__form"
    >
      <form
        className="form"
        onSubmit={[Function]}
      >
        <span
          className="text-input__span-container"
        >
          <input
            className="text-input"
            onChange={[Function]}
            placeholder="Enter Weight in kg"
            type="text"
            value=""
          />
          <div
            className="input-group__item input-group__item--date"
          >
            <withStyles(SingleDatePicker)
              anchorDirection="left"
              appendToBody={false}
              block={false}
              calendarInfoPosition="bottom"
              customCloseIcon={null}
              customInputIcon={null}
              date={"1969-12-31T23:00:00.000Z"}
              daySize={39}
              disableScroll={false}
              disabled={false}
              displayFormat={[Function]}
              enableOutsideDays={false}
              firstDayOfWeek={null}
              focused={null}
              hideKeyboardShortcutsPanel={false}
              horizontalMargin={0}
              horizontalMonthPadding={13}
              id="date_picker_weight"
              initialVisibleMonth={null}
              inputIconPosition="before"
              isDayBlocked={[Function]}
              isDayHighlighted={[Function]}
              isOutsideRange={[Function]}
              isRTL={false}
              keepFocusOnInput={false}
              keepOpenOnDateSelect={false}
              monthFormat="MMMM YYYY"
              navNext={null}
              navPrev={null}
              noBorder={false}
              numberOfMonths={1}
              onClose={[Function]}
              onDateChange={[Function]}
              onFocusChange={[Function]}
              onNextMonthClick={[Function]}
              onPrevMonthClick={[Function]}
              openDirection="down"
              orientation="horizontal"
              phrases={
                Object {
                  "calendarLabel": "Calendar",
                  "chooseAvailableDate": [Function],
                  "clearDate": "Clear Date",
                  "closeDatePicker": "Close",
                  "dateIsSelected": [Function],
                  "dateIsUnavailable": [Function],
                  "enterKey": "Enter key",
                  "escape": "Escape key",
                  "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
                  "homeEnd": "Home and end keys",
                  "jumpToNextMonth": "Move forward to switch to the next month.",
                  "jumpToPrevMonth": "Move backward to switch to the previous month.",
                  "keyboardBackwardNavigationInstructions": "Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
                  "keyboardForwardNavigationInstructions": "Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
                  "keyboardShortcuts": "Keyboard Shortcuts",
                  "leftArrowRightArrow": "Right and left arrow keys",
                  "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
                  "moveFocusByOneMonth": "Switch months.",
                  "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
                  "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
                  "openThisPanel": "Open this panel.",
                  "pageUpPageDown": "page up and page down keys",
                  "questionMark": "Question mark",
                  "returnFocusToInput": "Return to the date input field.",
                  "roleDescription": "datepicker",
                  "selectFocusedDate": "Select the date in focus.",
                  "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
                  "upArrowDownArrow": "up and down arrow keys",
                }
              }
              placeholder="Date"
              readOnly={false}
              regular={false}
              renderCalendarInfo={null}
              renderDayContents={null}
              renderMonthElement={null}
              renderMonthText={null}
              reopenPickerOnClearDate={false}
              required={false}
              screenReaderInputMessage=""
              showClearDate={false}
              showDefaultInputIcon={false}
              small={false}
              verticalHeight={null}
              verticalSpacing={22}
              weekDayFormat="dd"
              withFullScreenPortal={false}
              withPortal={true}
            />
          </div>
        </span>
        <button
          className="button button--form"
        >
          Update Weight
        </button>
      </form>
    </div>
  </div>
</div>
`;
